### PR TITLE
Add CLI

### PR DESCRIPTION
**Pull request by jacoblearned**
```
pcf -h
Usage: pcf [--version] [--help] <command> [<args>...]

  Particle Cloud Framework

Options:
  -v, --version  Show PCF version.
  -h, --help     Show this message and exit.

Commands:
  apply      Set a desired state and apply changes
  run        Set desired state to 'run' and apply changes
  stop       Set desired state to 'stopped' and apply changes
  terminate  Set desired state to 'terminated' and apply changes
```

#### Changes
Adds a [Click](http://click.palletsprojects.com/en/7.x/)-based CLI, `pcf`, with four commands (`apply`, `run`, `stop`, and `terminate`) for applying state changes to Particles and Quasiparticles
* CLI has autosuggestion for close typos
* Respects a user's `NO_COLOR` env var if set. See https://no-color.org

Adds a `test` Make task that will execute `pytest` and generate code coverage reports with uncovered line numbers.
* Adds a `.coveragerc` file to ignore the `pcf/test` directory in coverage reports

##### Overview
The CLI will by default look for the following config files in order of their precedence:
1. `pcf.json`
2. `pcf.yml`
3. `pcf.yaml`

Users can provide their own custom-named files with the `-f` or `--file` option.

These PCF config files can contain a single dict for a single particle definition or a list of particle definition dictionaries.

Example `pcf.yml` file:
```yaml
---
pcf_name: my_ec2
flavor: ec2_instance
aws_resource:
  custom_config:
    instance_name: my-ec2
    ...
```

The Particle/Quasiparticle lookup method searches for the corresponding class object that has the flavor given by the user set as a class variable since they are unique.

##### Commands
`pcf apply` is a general command that takes a `--state` or `-s` option (`running` as default), sets desired state, and runs `apply()` on the particle/quasiparticle class
* Example: `pcf apply -s terminated my_ec2`

`pcf run <pcf_name>` is a convenience command equivalent to `pcf apply --state running <pcf_name>`

`pcf stop <pcf_name>` is a convenience command equivalent to `pcf apply --state stopped <pcf_name>`

`pcf terminate <pcf_name>` is a convenience command equivalent to `pcf apply --state terminate <pcf_name>`

All four applying commands take a `-c` or `--cascade` flag to determine if state changes should cascade through family members as well as a `-q` or `--quiet` flag to prevent CLI output except when errors occur.

##### Tests
* Adds `conftest.py` with reusable pytest fixtures for a [CLIRunner](http://click.palletsprojects.com/en/7.x/api/#click.testing.CliRunner) and a convenient method to copy test config files into a temp directory.
* Adds tests for commands and all utility functions

To test the CLI, check out this branch, run `pip install -e .` and run `pcf --version` to get started.

### CODE REVIEW

**Comment by anovis** [comment]
looks really good. just a few things right now.

can we add a state command that returns the status of the particle/quasiparticle

after apply finishes

instead of` Successfully applied changes to ec2-example` maybe return the state

dont think commands should require `pcf_name` field. it could be optional if you only want to run one particle or a list of particles otherwise it will run everything in the json def

**Comment by anovis** [comment]
`Applying changes to ec2-example...` might want to display the pending state every interval... 
for apply can we add the max timeout parameter as well.

**Comment by anovis** [comment]
`Loading Particle/Quasiparticle flavor for ec2-example...`

should we display the actual flavor that it is loading?

File: pcf/cli/commands/run.py [hunk 1]
```diff
@@ -0,0 +1,41 @@
+""" Logic for pcf run command """
+
+import click
+from pcf.core import State
+from pcf.cli.utils import execute_applying_command
+
+
+@click.command(name="run", short_help="Set desired state to 'run' and apply changes")
```
Comment (anovis) [inline pcf/cli/commands/run.py]: should be state to `running`

File: pcf/cli/commands/terminate.py [hunk 2]
```diff
@@ -0,0 +1,45 @@
...
+@click.option(
+    "-f",
+    "--file",
+    "file_",
+    type=click.Path(dir_okay=False, resolve_path=True),
+    default="pcf.json",
+    show_default=True,
+    help="The JSON or YAML file defining your infrastructure configuration",
+)
+@click.option(
```
Comment (anovis) [inline pcf/cli/commands/terminate.py]: anyway to share these options so you dont have to rewrite them every-time
maybe using decorators?

https://stackoverflow.com/questions/40182157/python-click-shared-options-and-flags-between-commands

File: pcf/cli/utils.py [hunk 3]
```diff
@@ -0,0 +1,238 @@
...
+            continue
+
+        if recursive and is_pkg:
+            submodules += pkg_submodules(full_name)
+
+    return submodules
+
+
+def particle_class_from_flavor(flavor):
+    """ Return the class object of the given flavor (or None) by searching
```
Comment (anovis) [inline pcf/cli/utils.py]: would this work if the user has custom defined flavors not in the core... maybe env var with location or something in the config to specify.

File: pcf/cli/utils.py [hunk 4]
```diff
@@ -0,0 +1,238 @@
...
+        except ImportError:
+            continue
+
+        if recursive and is_pkg:
+            submodules += pkg_submodules(full_name)
+
+    return submodules
+
+
+def particle_class_from_flavor(flavor):
```
Comment (anovis) [inline pcf/cli/utils.py]: this should probably go into the core pcf utils. might add to the particle / quasiparticle class itself later on

File: pcf/cli/utils.py [hunk 5]
```diff
@@ -0,0 +1,238 @@
...
+            "Setting desired state of {0} to {1}...".format(pcf_name, desired_state),
+            fg=color("blue"),
+        )
+
+    particle.set_desired_state(getattr(State, desired_state))
+
+    if not quiet:
+        click.secho("Applying changes to {0}...".format(pcf_name), fg=color("blue"))
+
+    particle.apply(cascade=cascade)
```
Comment (anovis) [inline pcf/cli/utils.py]: add max timeout params here as well.

**Review by anovis — CHANGES_REQUESTED** [review]

File: pcf/cli/commands/terminate.py (hunk 2 above)
Comment (jacoblearned) [reply pcf/cli/commands/terminate.py]: Yeah that's a good call.

File: pcf/cli/utils.py (hunk 3 above)
Comment (jacoblearned) [reply pcf/cli/utils.py]: Not this function. I wasn't aware of a formal process for building custom Particles external to the package. Is this documented anywhere?

**Comment by jacoblearned** [comment]
@anovis Since this PR is quite large, I was planning on adding the state command in a subsequent PR. I agree with the preference to show the state after apply finishes as well and will include that too.

**Comment by jacoblearned** [comment]
> `Applying changes to ec2-example...` might want to display the pending state every interval...
> for apply can we add the max timeout parameter as well.

What interval are you referring to?

**Comment by jacoblearned** [comment]
@anovis I pushed some changes for the suggestions. Let me know what you think whenever you have the time 👍

**Comment by anovis** [comment]
there is no set interval now so it would be up to you for what makes sense. this was just a suggestion, but some particles can take awhile so it might be good for the user to see that it is in a pending state.

File: pcf/cli/commands/__init__.py [hunk 6]
```diff
@@ -0,0 +1,31 @@
+import click
+
+COMMON_OPTIONS = [
```
Comment (anovis) [inline pcf/cli/commands/__init__.py]: these are just common options for the apply commands. so maybe change to apply_options or common_apply_options.

File: pcf/cli/commands/__init__.py [hunk 7]
```diff
@@ -0,0 +1,31 @@
...
+        "-c",
+        "--cascade",
+        is_flag=True,
+        help="Apply state transitions to all family members",
+    ),
+    click.option(
+        "-t",
+        "--timeout",
+        type=int,
+        help="The maximum number of seconds to wait before a timeout error is thrown",
```
Comment (anovis) [inline pcf/cli/commands/__init__.py:29]: default is None

File: pcf/cli/commands/stop.py [hunk 8]
```diff
@@ -0,0 +1,29 @@
+""" Logic for pcf stop command """
+
+import click
+from pcf.core import State
+from pcf.cli.utils import execute_applying_command, click_options
+from pcf.cli.commands import COMMON_OPTIONS
+
+
+@click.command(
+    name="stop", short_help="Set desired state to 'stopped' and apply changes"
```
Comment (anovis) [inline pcf/cli/commands/stop.py]: just apply not `apply changes`

File: pcf/cli/commands/terminate.py [hunk 9]
```diff
@@ -0,0 +1,29 @@
+""" Logic for pcf stop command """
+
+import click
+from pcf.core import State
+from pcf.cli.utils import execute_applying_command, click_options
+from pcf.cli.commands import COMMON_OPTIONS
+
+
+@click.command(
+    name="terminate", short_help="Set desired state to 'terminated' and apply changes"
```
Comment (anovis) [inline pcf/cli/commands/terminate.py]: just apply not `apply changes`

File: pcf/cli/commands/terminate.py [hunk 10]
```diff
@@ -0,0 +1,29 @@
...
+)
+@click_options(COMMON_OPTIONS)
+@click.argument("pcf_name", required=False)
+@click.pass_context
+def terminate(ctx, pcf_name, cascade, quiet, file_, timeout):
+    """ Set desired state to 'terminated' and apply changes
```
Comment (anovis) [inline pcf/cli/commands/terminate.py]: just apply not `apply changes`

**Review by anovis — CHANGES_REQUESTED** [review]

File: pcf/cli/utils.py (hunk 3 above)
Comment (anovis) [reply pcf/cli/utils.py]: https://capitalone.github.io/Particle-Cloud-Framework/docs/build/html/sections/quickstartdeveloper.html

you would use the same way, but would import from a different path

File: pcf/cli/commands/__init__.py (hunk 7 above)
Comment (jacoblearned) [reply pcf/cli/commands/__init__.py:29]: If the flag isn't passed by the user it will be `None` by default.

**Comment by jacoblearned** [comment]
@anovis I added a progress bar for when `apply` runs on multiple particles.

File: pcf/cli/utils.py (hunk 3 above)
Comment (jacoblearned) [reply pcf/cli/utils.py]: That doc recommends pull requests into this package to integrate custom-built Particles and Quasiparticles. Are you suggesting that the CLI should support an optional field in the config file per particle deployment in which users have to specify Python module path strings for this function to look for like this:

```yaml
pcf_name: my_custom_particle
flavor: my_custom_particle_flavor
external_package: my_module.my_custom_particle
aws_resources:
   ...
```

And then the CLI strip out the external_package field before passing the config to the Particle constructor if it finds the right class in `my_module.my_custom_particle`?

**Review by anovis — APPROVED** [review]

**Comment by anovis** [comment]
is there a better way to install the cli for a user than running `pip install -e .`

**Comment by anovis** [comment]
can you add a readme page or add a section to the docs? even just the screenshot you have in the pr works!

**Comment by jacoblearned** [comment]
> is there a better way to install the cli for a user than running `pip install -e .`

`pip install -e .` is just for testing the CLI in a development environment. It creates an editable install as defined by the root `setup.py`. 

A regular user would just run `pip install pcf` and then have access to the `pcf` CLI because it is defined as a [`console_script` entrypoint in `setup.py`](https://github.com/capitalone/Particle-Cloud-Framework/blob/add-apply-command/setup.py#L35-L38)

**Comment by davidyum** [comment]
awesome thanks Jacob! Will review and try this out today

File: pcf/__init__.py [hunk 11]
```diff
@@ -0,0 +1,3 @@
+import os
+
+VERSION = os.environ.get('PCF_TAG', '0.1.1')
```
Comment (davidyum) [inline pcf/__init__.py]: I think we are at pcf 0.3.2 now.

File: pcf/cli/commands/apply.py [hunk 12]
```diff
@@ -0,0 +1,34 @@
...
+from pcf.cli.utils import execute_applying_command, click_options
+from pcf.cli.commands import COMMON_APPLY_OPTIONS
+
+
+@click.command(name="apply", short_help="Set a desired state and apply")
+@click.option(
+    "-s",
+    "--state",
+    type=click.Choice(["running", "stopped", "terminated"], case_sensitive=False),
+    default="running",
```
Comment (davidyum) [inline pcf/cli/commands/apply.py:13]: we might want to force them to specify desired state instead of defaulting to running for consistency.

**Review by davidyum — COMMENTED** [review]
looks good! Tested it with aws/gcp particles and quasiparticles and all worked smoothly.

**Comment by davidyum** [comment]
Anything like a README or how to get started guide for users would be helpful. Almost just like the first comment you have at the top :)

**Comment by davidyum** [comment]
@jacoblearned theres a merge conflict with pcf/particle/gcp/compute/compute.py because we renamed but I think we are good to merge once we have that fixed.

**Comment by jacoblearned** [comment]
@davidyum Resolved the conflict. Thanks for reminding me 👍

**Review by davidyum — APPROVED** [review]

**Pull request merged**